<table align="center" width=100%>
    <tr>
        <td width="20%">
            <img src="in_class.png">
        </td>
        <td>
            <div align="center">
                <font color="#21618C" size=8px>
                    <b> Inclass-Lab<br>(Week 2)
                    </b>
                </font>
            </div>
        </td>
        <td width="25%">
            <img src="NEW_GL_Logo.png">
        </td>
    </tr>
</table>

##  Table of Content

1. **[Conditional Statements](#cond)**
2. **[Loops](#loop)**
3. **[List Comprehensions](#comp)**

### Let's begin with some hands-on practice exercises

<a id="cond"> </a>
## 1. Conditional Statements

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>1. Write a code to print the maximum of two numbers (use if-else)</b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [2]:
# type your code here
num1 = 100; num2 = 150

if num1 > num2:
    print(f"Num1 is the maximum number. {num1}")
elif num1 == num2:
    print("Both number are same")
else:
    print(f"Num2 is the maximum number. {num2}")

Num2 is the maximum number. 150


<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>2. Write a program to check whether a number is positive, negative or zero (use if-elif) </b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [ ]:
# type your code here


<a id= "loop"> </a>
## 2. Loops

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>3. Write a program for generating a fibonacci series starting with 0 and 1 for the next 10 values using a while loop. </b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [7]:
# type your code here
# [0,1,1,2,3,5,8,13,21,34,55,89]

a = 0; b = 1
fibo = [a,b]
i = 1
while i < 11:
    c = a + b
    a = b
    b = c
    i = i + 1
    fibo.append(b)
    
print(fibo)

[0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89]


In [10]:
# type your code here
# [0,1,1,2,3,5,8,13,21,34,55,89]

a = 0; b = 1
fibo = [a,b]
i = 1
while i <= 10:
    c = a + b
    a = b
    b = c
    i = i + 1
    fibo.append(b)
    
print(fibo)

[0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89]


### Rishab

In [13]:
a = 0 ; b = 1
fibo = [a,b]
i =1
while i <11:
    c = a+b
    a = b
    b = c
    i = i+1 
fibo.append(b)

In [14]:
fibo

[0, 1, 89]

In [17]:
a = 0 ; b = 1
fibo = [a,b]
i =1
while i <11:
    c = a+b
    a = b
    b = c
    i = i+1 
    fibo.append(b)
print(fibo)

[0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89]


<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>4. Write a code to find factorial of a number (take the input from user). </b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [12]:
# type your code here
num = int(input("Enter a number: "))
fact = 1
if num < 0:
    print("Sorry, factorial does not exist for negative number.")
elif num == 0:
    print("The factorial of 0 is 1.")
else:
    for i in range(1, num + 1):
        fact = fact * i
    print(f"The factorial of {num} is {fact}")

Enter a number: 5
The factorial of 5 is 120


<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>5. Write a program to check whether a number is prime or not (take the input from user).</b>
                </font>
            </div>
        </td>
    </tr>
</table>

# type your code here
# already solve
# Strongly watch this:

# develop a game using python programming based on while loop and nested structure
https://youtu.be/VauAiVixICU?si=N9t7rTNZMKppsd0a


<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>6. Write a program to check whether two numbers are amicable or not (take the input from user).</b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [18]:
# type your code here
num1 = int(input("Enter the first number: "))
num2 = int(input("Enter the second number: "))
sum_factor_1 = 0
sum_factor_2 = 0

for i in range(1, num1):
    if num1 % i == 0:
        sum_factor_1 = sum_factor_1 + i
        print(i)
        
for j in range(1, num2):
    if num2 % j == 0:
        sum_factor_2 = sum_factor_2 + j
        print(j)
        
if (sum_factor_1 == num2 and sum_factor_2 == num1):
    print(f"the numbers {num1} and {num2} are amicable.")
else:
    print(f"the numbers {num1} and {num2} are not amicable.")
    
print(sum_factor_1, sum_factor_2)

Enter the first number: 220
Enter the second number: 284
1
2
4
5
10
11
20
22
44
55
110
1
2
4
71
142
the numbers 220 and 284 are amicable.
284 220


<a id= "lc"> </a>
## 2. List Comprehension

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>7. Use list comprehension to find the used vowels in a given sentence and/or word (take the input from user). </b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [19]:
s = input("Enter a word or phrase: ")
for vowels in s:
    if vowels.lower() in ["a",'e','i','o','u']:
        print(vowels)

Enter a word or phrase: We are learning Python programming.
e
a
e
e
a
i
o
o
a
i


In [24]:
s = input("Enter a word or phrase: ")
[vowels for vowels in s if vowels.lower() in ["a",'e','i','o','u'] ]

Enter a word or phrase: great learning


['e', 'a', 'e', 'a', 'i']

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>8. Create a dictionary and access it values using a condition on its key. The data is given below. Let the condidtion on the key is that it should be a even number.</b>
                </font>
            </div>
       </td>
    </tr>
</table>


<table>
<tr>
    <th>Key</th>
    <td>1</td>
    <td>2</td>
    <td>3</td>
    <td>4</td>
    <td>5</td>
</tr>
<tr>
    <th>Name</th>
    <td>Aman</td>
    <td>Mohit</td>
    <td>Gauri</td>
    <td>Imran</td>
    <td>Roma</td
     </tr>
<tr>
    <th>Marks</th>
    <td>24</td>
    <td>25</td>
    <td>26</td>
    <td>24</td>
    <td>27</td>
   </tr>
</table>

In [27]:
# type your code here
data = {1:['Aman',24], 2 : ['Mohit',25], 3:["Gauri",26],4 :["Imran",24],5:['Roma',27]}
data

{1: ['Aman', 24],
 2: ['Mohit', 25],
 3: ['Gauri', 26],
 4: ['Imran', 24],
 5: ['Roma', 27]}

In [31]:
for key, values in data.items():
    if key % 2 == 0:
        print({key:values})

{2: ['Mohit', 25]}
{4: ['Imran', 24]}


In [33]:
{key:values for key, values in data.items() if key % 2 == 0}

{2: ['Mohit', 25], 4: ['Imran', 24]}

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>9. Use list comprehension to print table of 11 to 20 as shown below.</b>
                </font>
            </div>
        </td>
    </tr>
</table>

<table>
<tr>
    <td>11</td><td>22</td><td>33</td><td>44</td><td>55</td><td>66</td><td>77</td><td>88</td><td>99</td><td>110</td>
</tr>
    
<tr>
    <td>12</td><td>24</td><td>36</td><td>48</td><td>60</td><td>72</td><td>84</td><td>96</td><td>108</td><td>120</td>
</tr>

<tr>
    <td>13</td><td>36</td><td>39</td><td>52</td><td>65</td><td>78</td><td>91</td><td>104</td><td>117</td><td>130</td>
</tr>

<tr>
     <td>14</td><td>28</td><td>42</td><td>56</td><td>70</td> <td>84</td><td>98</td><td>112</td><td>126</td><td>140</td>
</tr>

<tr>
     <td>15</td><td>30</td><td>45</td><td>60</td><td>75</td><td>90</td><td>105</td><td>120</td><td>135</td><td>150</td>
</tr>

<tr>
     <td>16</td><td>32</td><td>48</td><td>64</td><td>80</td><td>96</td><td>112</td><td>128</td><td>144</td><td>160</td>
</tr>

<tr>
     <td>17</td><td>34</td><td>51</td><td>68</td><td>85</td><td>102</td><td>119</td><td>136</td><td>153</td><td>170</td>
</tr>

<tr>
     <td>18</td><td>36</td><td>54</td><td>72</td><td>90</td> <td>108</td><td>126</td><td>144</td><td>162</td><td>180</td>
</tr>

<tr>
 <td>19</td><td>38</td><td>57</td><td>76</td><td>95</td><td>114</td><td>133</td><td>152</td><td>171</td><td>190</td>
</tr>

<tr>
     <td>20</td><td>40</td><td>60</td><td>80</td><td>100</td><td>120</td><td>140</td><td>160</td><td>180</td><td>200</td>
<tr>    
    
</table>

In [37]:
# type your code here
for i in range(11,21):
    print([i * j for j in range(1,11)])


[11, 22, 33, 44, 55, 66, 77, 88, 99, 110]
[12, 24, 36, 48, 60, 72, 84, 96, 108, 120]
[13, 26, 39, 52, 65, 78, 91, 104, 117, 130]
[14, 28, 42, 56, 70, 84, 98, 112, 126, 140]
[15, 30, 45, 60, 75, 90, 105, 120, 135, 150]
[16, 32, 48, 64, 80, 96, 112, 128, 144, 160]
[17, 34, 51, 68, 85, 102, 119, 136, 153, 170]
[18, 36, 54, 72, 90, 108, 126, 144, 162, 180]
[19, 38, 57, 76, 95, 114, 133, 152, 171, 190]
[20, 40, 60, 80, 100, 120, 140, 160, 180, 200]


In [38]:
[[i * j for j in range(1,11)] for i in range(11,21) ]

[[11, 22, 33, 44, 55, 66, 77, 88, 99, 110],
 [12, 24, 36, 48, 60, 72, 84, 96, 108, 120],
 [13, 26, 39, 52, 65, 78, 91, 104, 117, 130],
 [14, 28, 42, 56, 70, 84, 98, 112, 126, 140],
 [15, 30, 45, 60, 75, 90, 105, 120, 135, 150],
 [16, 32, 48, 64, 80, 96, 112, 128, 144, 160],
 [17, 34, 51, 68, 85, 102, 119, 136, 153, 170],
 [18, 36, 54, 72, 90, 108, 126, 144, 162, 180],
 [19, 38, 57, 76, 95, 114, 133, 152, 171, 190],
 [20, 40, 60, 80, 100, 120, 140, 160, 180, 200]]

In [41]:
for func in dir(list):
    if "__" not in func:
        print(func)

append
clear
copy
count
extend
index
insert
pop
remove
reverse
sort


In [43]:
print([func for func in dir(list) if "__" not in func])

['append', 'clear', 'copy', 'count', 'extend', 'index', 'insert', 'pop', 'remove', 'reverse', 'sort']


In [44]:
print([func for func in dir(set) if "__" not in func])
print()
print([func for func in dir(tuple) if "__" not in func])
print()
print([func for func in dir(dict) if "__" not in func])

['add', 'clear', 'copy', 'difference', 'difference_update', 'discard', 'intersection', 'intersection_update', 'isdisjoint', 'issubset', 'issuperset', 'pop', 'remove', 'symmetric_difference', 'symmetric_difference_update', 'union', 'update']

['count', 'index']

['clear', 'copy', 'fromkeys', 'get', 'items', 'keys', 'pop', 'popitem', 'setdefault', 'update', 'values']


### List comprehension:

https://www.youtube.com/watch?v=NCkoq5DIRHc&list=PLWuFHho1zKhWb-f-SJAMUCK--f8PJlG46&index=27

In [45]:
l=[[1 ,2, 3], [4, 5, 6], [7, 8, 9]]
[[row[i] for row in l] for i in range(3)]
    

[[1, 4, 7], [2, 5, 8], [3, 6, 9]]

In [48]:
for i in range(3):
    print([row[i] for row in l])


[1, 4, 7]
[2, 5, 8]
[3, 6, 9]


In [49]:
[[row[i] for row in l] for i in range(3)]

[[1, 4, 7], [2, 5, 8], [3, 6, 9]]

In [50]:
[i for i in range(20) if i % 2 == 0]

[0, 2, 4, 6, 8, 10, 12, 14, 16, 18]

In [51]:
[i for i in range(20) if i % 2 != 0]

[1, 3, 5, 7, 9, 11, 13, 15, 17, 19]

In [ ]:
[m for m in mobile if (str(m).isnumeric() and len(str(m) == 10)) 
                         and (str(m)[:2] == "98" or str(m)[:2] == "99")]

In [52]:
["3PrintYourLoop_Varible 1applyYourLoop 2ApplyYourcondition_logicalOperator"]

['3PrintYourLoop_Varible 1applyYourLoop 2ApplyYourcondition_logicalOperator']